packages/lesswrong/server/languageModels/languageModelIntegration.ts

```diff
@@ -0,0 +1,307 @@
+ * shorten truncatableVariable to fit.
+ *
+ * This is NOT safe for SQL, HTML rendering, or anything else that's sensitive
+ * to quoting. It is intended only for use with language-model prompting.
+ */
+export function substituteIntoTemplate({template, variables, maxLengthTokens, truncatableVariable}: {
+  template: LanguageModelTemplate,
+  variables: Record<string,string>
+  maxLengthTokens?: number,
+  truncatableVariable?: string,
+}): string {
+  let withVarsSubstituted = template.body;
+  
+  // Substitute everything except the truncatable variable
+  for (let key of Object.keys(variables)) {
+    if (key !== truncatableVariable || !maxLengthTokens)
+      withVarsSubstituted = withVarsSubstituted.replace(new RegExp("\\${"+key+"}", "g"), variables[key]);
+  }
+  
+  if (maxLengthTokens && truncatableVariable) {
+    const withVarsSubstitutedAndTruncVarRemoved = withVarsSubstituted.replace(new RegExp("\\${"+truncatableVariable+"}", "g"), "");
+    const tokensSpent = countGptTokens(withVarsSubstitutedAndTruncVarRemoved);
+    const tokensAvailable = maxLengthTokens - tokensSpent;
+    const truncatedVar = truncateByTokenCount(variables[truncatableVariable]||"", tokensAvailable);
+    withVarsSubstituted = withVarsSubstituted.replace(new RegExp("\\${"+truncatableVariable+"}", "g"), truncatedVar);
+  }
+  
+  return withVarsSubstituted;
+}
+
+function countGptTokens(str: string): number {
+  if (!str) return 0;
+  try {
+    return gpt3encode(str).length;
+  } catch(e) {
+    return str.length;
+  }
+}
+
+/**
+ * Truncate a string to a given length, measured in GPT-3 tokens (which only
+ * approximately line up with character counts). Uses countGptTokens (which uses
+ * gpt-3-encoder) plus binary search. (This could be made faster by instead
+ * doing an encode-then-decode round trip, but that raises potential bugs if
+ * some character sequences don't roundtrip.)
+ */
+function truncateByTokenCount(str: string, tokens: number): string {
+  if (!str || !str.length)
+    return "";
+  if (str.length < tokens)
+    return str;
+  
+  // First try an encode-then-decode roundtrip
+  try {
+    const encoded = gpt3encode(str);
+    
+    if (encoded.length <= tokens) return str;
+    const redecoded = gpt3decode(take(encoded,tokens));
+    if (redecoded === str.substring(0,redecoded.length)) {
+      return redecoded;
+    } else {
+      // eslint-disable-next-line no-console
+      console.log(`GPT-3 encoding did not roundtrip: ${JSON.stringify(str)}`);
+    }
+  } catch {
+    console.log(`Could not encode string for truncation length estimate: ${JSON.stringify(str)}`); //eslint-disable-line no-console
+  }
+  
+  // If that didn't work, binary-search string truncations to find one that has the right token count
+  let low=0, high=str.length, mid=(low+high)/2|0;
+  
+  while (high>low) {
+    if (countGptTokens(str.substring(0,mid)) > tokens) {
+      high = mid;
+    } else {
+      low = mid+1;
+    }
+    mid=(low+high)/2|0;
```

Comment:
Is the bitwise operation just intended to act as `truncate`?  A comment would probably be good here

Reply:
Commented